File: smart_gym_planner_web/src/utils/workoutGenerator.js
import { allExercises, warmupExercises, stretchingExercises, cardioExercises } from '../data/exerciseData';

export function generateWorkoutPlan(user) {
  const hasEquipment = user.equipment && !user.equipment.includes('none') && user.equipment.length > 0;
  const splits = getSplitsForDays(user.workoutDaysPerWeek);

  const days = Array.from({ length: user.workoutDaysPerWeek }, (_, i) => {
    const split = splits[i];
    const exercises = getExercisesForSplit(split, user.equipment || [], user.experienceLevel, hasEquipment);
    
    // Calculate total sets across main exercises
    const totalSets = exercises.reduce((sum, e) => sum + e.sets, 0);
    // Rough calorie burn estimate: (exercises * sets * 8) + 120 (cardio/warmup)
    const estimatedCalories = (exercises.length * totalSets * 8) + 120;

    return {
      dayName: `Day ${i + 1}`,
      splitName: split,
      exercises,
      warmup: warmupExercises,
      stretching: stretchingExercises,
      cardio: cardioExercises,
      isCompleted: false,
      estimatedCalories,
    };
  });

  return {
    days,
    generatedFor: user.experienceLevel,
  };
}

function getSplitsForDays(days) {
  switch (days) {
    case 5:
      return [
        'Chest + Triceps',
        'Back + Biceps',
        'Legs',
        'Shoulders + Core',
        'Full Body',
      ];
    case 6:
      return [
        'Chest + Triceps',
        'Back + Biceps',
        'Shoulders + Forearms',
        'Chest + Triceps',
        'Back + Biceps',
        'Legs + Shoulders',
      ];
    case 7:
    default:
      return [
        'Chest + Triceps',
        'Back + Biceps',
        'Legs + Shoulders',
        'Shoulders + Forearms',
        'Chest + Triceps',
        'Back + Biceps',
        'Legs + Shoulders',
      ];
  }
}

function getExercisesForSplit(split, equipmentList, level, hasEquipment) {
  const muscles = getMusclesForSplit(split);

  const filtered = allExercises.filter(e => {
    // Check muscle group
    if (!muscles.includes(e.muscleGroup)) return false;
    
    // Check equipment
    if (!hasEquipment) {
      return e.requiredEquipment.length === 0;
    }
    if (e.requiredEquipment.length === 0) return true;
    return e.requiredEquipment.some(eq => equipmentList.includes(eq));
  }).filter(e => {
    // Check difficulty level
    if (level === 'beginner') return e.difficulty === 'beginner';
    if (level === 'intermediate') return e.difficulty !== 'expert';
    return true; // expert sees all
  });

  const count = level === 'beginner' ? 4 : (level === 'intermediate' ? 5 : 6);
  return filtered.slice(0, count);
}

function getMusclesForSplit(split) {
  switch (split) {
    case 'Chest + Triceps':
      return ['chest', 'triceps'];
    case 'Back + Biceps':
      return ['back', 'biceps'];
    case 'Legs':
      return ['legs'];
    case 'Legs + Shoulders':
      return ['legs', 'shoulders'];
    case 'Shoulders + Forearms':
      return ['shoulders', 'forearms'];
    case 'Shoulders + Core':
      return ['shoulders', 'core'];
    case 'Full Body':
      return ['chest', 'back', 'legs', 'shoulders'];
    default:
      return [];
  }
}

File: smart_gym_planner/lib/services/workout_generator.dart
import '../models/exercise_model.dart';
import '../models/user_model.dart';
import 'exercise_data.dart';

/// Generates a weekly workout plan based on user profile
class WorkoutGenerator {
  static WorkoutPlan generate(UserModel user) {
    final bool hasEquipment =
        !user.equipment.contains('none') && user.equipment.isNotEmpty;
    final splits = _getSplits(user.workoutDaysPerWeek);

    final days = List.generate(user.workoutDaysPerWeek, (i) {
      final split = splits[i];
      final exercises = _exercisesForSplit(
          split, user.equipment, user.experienceLevel, hasEquipment);
      return WorkoutDay(
        dayName: 'Day ${i + 1}',
        splitName: split,
        exercises: exercises,
        warmup: ExerciseData.warmupExercises,
        stretching: ExerciseData.stretchingExercises,
        cardio: ExerciseData.cardioExercises, // cardio every day
      );
    });

    return WorkoutPlan(days: days, generatedFor: user.experienceLevel);
  }

  /// Returns the correct split schedule for 5, 6, or 7 days
  static List<String> _getSplits(int days) {
    switch (days) {
      case 5:
        return [
          'Chest + Triceps',
          'Back + Biceps',
          'Legs',
          'Shoulders + Core',
          'Full Body',
        ];
      case 6:
        return [
          'Chest + Triceps',
          'Back + Biceps',
          'Shoulders + Forearms',
          'Chest + Triceps',
          'Back + Biceps',
          'Legs + Shoulders',
        ];
      case 7:
      default:
        return [
          'Chest + Triceps',
          'Back + Biceps',
          'Legs + Shoulders',
          'Shoulders + Forearms',
          'Chest + Triceps',
          'Back + Biceps',
          'Legs + Shoulders',
        ];
    }
  }

  static List<ExerciseModel> _exercisesForSplit(
    String split,
    List<String> equipment,
    String level,
    bool hasEquipment,
  ) {
    final muscles = _musclesForSplit(split);

    var filtered = ExerciseData.allExercises.where((e) {
      if (!muscles.contains(e.muscleGroup)) return false;
      if (!hasEquipment) return e.requiredEquipment.isEmpty;
      if (e.requiredEquipment.isEmpty) return true;
      return e.requiredEquipment.any((eq) => equipment.contains(eq));
    }).where((e) {
      if (level == 'beginner') return e.difficulty == 'beginner';
      if (level == 'intermediate') return e.difficulty != 'expert';
      return true; // expert sees all
    }).toList();

    final count = level == 'beginner' ? 4 : (level == 'intermediate' ? 5 : 6);
    return filtered.take(count).toList();
  }

  static List<String> _musclesForSplit(String split) {
    switch (split) {
      case 'Chest + Triceps':
        return ['chest', 'triceps'];
      case 'Back + Biceps':
        return ['back', 'biceps'];
      case 'Legs':
        return ['legs'];
      case 'Legs + Shoulders':
        return ['legs', 'shoulders'];
      case 'Shoulders + Forearms':
        return ['shoulders', 'forearms'];
      case 'Shoulders + Core':
        return ['shoulders', 'core'];
      case 'Full Body':
        return ['chest', 'back', 'legs', 'shoulders'];
      default:
        return [];
    }
  }
}
